### PR TITLE
Introduce Javascript Routing

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
- Introduces FOSJS Routing Bundle
 - Replace most of the routes with actual Route Generation
 - Replace the other by adding at least the Base-Path, wasn't sure how to handle those properly
 - Replace all static route uses in PHP and Templates

This requires precise review and testing. I already tested most parts, but I am pretty sure I missed something.

### CODE REVIEW

**Comment by dpfaffenbauer** [comment]
With this change, It might be also possible now to change the admin route from /admin to something else, right?

**Comment by brusch** [comment]
> With this change, It might be also possible now to change the admin route from /admin to something else, right?

@dpfaffenbauer in theory yes, but there are more things to consider, e.g. Firewall, PimcoreContextResolver, ...

**Comment by brusch** [comment]
First of all, thanks a lot for your PR! 
After the first quick review this looks quite nice in general. 
However there's one thing that should be definitely considered before we start a detailed review/testing: the names of the routes should be configured in the `@Route` annotation to make the controller/route lookup more comfortable. We can of course just reuse the auto-generated names for that purpose.

**Comment by dpfaffenbauer** [comment]
Yes, I understand that. But: I think it is easier to make that change once we actually use route-names in javascript. As it is way easier to change route names and Search/Replace throughout the code to find usages of said routes.

**Comment by brusch** [comment]
Yes, but that should be possible in this PR/branch, right, JS is done there already 😉

**Comment by dpfaffenbauer** [comment]
So should I change all route names now? I would suggest to first check this PR, then to change the route names.

**Comment by brusch** [comment]
Yes, please. 
I've roughly checked that PR already, so far I haven't seen any major showstoppers, so you can go ahead 👍😉

**Comment by dpfaffenbauer** [comment]
alright, will take some time though. 

Any schema I should apply to route names?

**Comment by brusch** [comment]
Yes, of course 😉
Regarding the schema ... in terms of BC it probably would make sense to keep the auto-generated names, just in case someone's code relies on the current route names.

**Comment by dpfaffenbauer** [comment]
Alright, makes things a bit easier then ;)

**Comment by dpfaffenbauer** [comment]
@brusch done

**Comment by dpfaffenbauer** [comment]
> > With this change, It might be also possible now to change the admin route from /admin to something else, right?
> 
> @dpfaffenbauer in theory yes, but there are more things to consider, e.g. Firewall, PimcoreContextResolver, ...

we'll find a solution for those too ;)

**Comment by dpfaffenbauer** [comment]
@brusch Maybe you could add this PR to the April Contributions thing you make, this really needs lots of testing and review and this might help here. WDYT?

**Comment by brusch** [comment]
@dpfaffenbauer would be a bit hard to measure, but maybe we find another way 😊

**Comment by dvesh3** [comment]
@dpfaffenbauer can you please fix the tests?  thanks!

**Comment by dpfaffenbauer** [comment]
sure, should be done

**Comment by dpfaffenbauer** [comment]
@brusch did you get time to check this yet?

**Comment by brusch** [comment]
@dpfaffenbauer not yet, sorry 😐

**Comment by msphn** [comment]
I don't know, but that concept feels wrong, it's coupling the admin closer to the actual backend. Wouldn't be a good openapi docs better? There are client generators for openapi (for js, php everything) and a nice documentation would be a side effect.

**Comment by dpfaffenbauer** [comment]
Everything already is coupled pretty tightly. I don't see any issue using route names instead of actual paths. It also comes with the advantage of using the same route-names already existing in Symfony. Changing the route itself doesn't affect any JS side. It also comes with the advantage of respecting base-paths like `/app_test.php/admin`.

**Comment by dpfaffenbauer** [comment]
@brusch anything I can help to speed this up?

**Comment by brusch** [comment]
@dpfaffenbauer Divesh is currently reviewing this PR. @dvesh3 if there's anything @dpfaffenbauer can help, please share directly with him 😊
Thanks guys!

**Comment by dpfaffenbauer** [comment]
@dvesh3 just so you know, I just merged master into this again and resolved the conflicts

**Comment by dvesh3** [comment]
Hey @dpfaffenbauer thanks for your help! 🙏  I'm in middle of first round of review and will request changes soon. So you can help us by providing fixes asap so that we finish the review and merge it 😉

**Comment by dpfaffenbauer** [comment]
@dvesh3 depends on how fast you can provide feedback ;)

**Comment by dvesh3** [comment]
@dpfaffenbauer i'm still not done yet but maybe you can have a look at request changes so far. Thanks!

File: bundles/AdminBundle/Controller/Admin/DataObject/ClassificationstoreController.php [hunk 1]
```diff
@@ -477,7 +477,7 @@ public function groupsActionGet(Request $request)
     }
 
     /**
-     * @Route("/groups", methods={"POST", "PUT"})
+     * @Route("/groups", name="pimcore_admin_dataobject_classificationstore_groupsactionget", methods={"POST", "PUT"})
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/DataObject/ClassificationstoreController.php:480]: the same route gets used for GET, POST and PUT in JS, so can't rename them differently

File: bundles/AdminBundle/Controller/Admin/Document/DocumentControllerBase.php [hunk 2]
```diff
@@ -161,8 +161,6 @@ protected function addTranslationsData(Model\Document $document, array &$data)
     }
 
     /**
-     * @Route("/save-to-session", methods={"POST"})
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/Document/DocumentControllerBase.php:164]: don't understand what you mean? It has been removed?!

[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/Document/DocumentControllerBase.php:164]: ah, got it, you mean the use Route. Gotcha

File: bundles/AdminBundle/Controller/Admin/ElementController.php [hunk 3]
```diff
@@ -34,7 +34,7 @@
 class ElementController extends AdminController
 {
     /**
-     * @Route("/element/lock-element", methods={"PUT"})
+     * @Route("/element/lock-element", name="pimcore_admin_element_lockelement", methods={"PUT"})
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/ElementController.php:37]: Honestly, that has nothing todo with this PR and I would prefer to change as little as possible with route and route names. I can provide a follow up PR.

File: bundles/AdminBundle/Controller/Admin/ElementControllerBase.php [hunk 4]
```diff
@@ -173,7 +169,7 @@ public function deleteInfoAction(Request $request)
                         $deleteObjectsPerRequest = 5;
                         for ($i = 0; $i < ceil($childs / $deleteObjectsPerRequest); $i++) {
                             $deleteJobs[] = [[
-                                'url' => '/admin/' . $type . '/delete',
+                                'url' => $this->get('router')->getContext()->getBaseUrl() . '/admin/' . $type . '/delete',
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/ElementControllerBase.php:172]: It is, but you'd have to map $type to the routes.

File: bundles/AdminBundle/Controller/Admin/TranslationController.php [hunk 5]
```diff
@@ -764,7 +764,7 @@ public function contentExportJobsAction(Request $request)
         $elements = array_chunk($elements, $elementsPerJob);
         foreach ($elements as $chunk) {
             $jobs[] = [[
-                'url' => '/admin/translation/' . $type . '-export',
+                'url' => $this->get('router')->getContext()->getBaseUrl() . '/admin/translation/' . $type . '-export',
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/TranslationController.php:767]: Yeah no, would refactor that anyway to make it more dynamic. With this it is currently possible to add custom export types. But I would introduce a registry with possible exports and use one route for all of them. Can be done in a future PR.

File: bundles/AdminBundle/Controller/Admin/UserController.php [hunk 6]
```diff
@@ -35,7 +35,7 @@
 class UserController extends AdminController implements EventedControllerInterface
 {
     /**
-     * @Route("/user/tree-get-childs-by-id", methods={"GET"})
+     * @Route("/user/tree-get-childs-by-id", name="pimcore_admin_user_treegetchildsbyid", methods={"GET"})
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/UserController.php:38]: As above, not part of this PR

File: bundles/AdminBundle/Resources/config/pimcore/routing.yml [hunk 7]
```diff
@@ -2,26 +2,36 @@ _pimcore_admin:
     resource: "@PimcoreAdminBundle/Controller/Admin/"
     type:     annotation
     prefix:   /admin
+    options:
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Resources/config/pimcore/routing.yml:5]: Ecommerce framework has not been tackled at all yet, so not necessary. Should also be done in a separate PR. This only focues on Core Pimcore. To keep complexity down, I can provide a separate PR for ecommerce

File: bundles/AdminBundle/Resources/public/js/pimcore/document/document.js [hunk 8]
```diff
@@ -14,12 +14,10 @@
 pimcore.registerNS("pimcore.document.document");
 pimcore.document.document = Class.create(pimcore.element.abstract, {
 
-    urlprefix: "/admin/",
-
     getData: function () {
         var options = this.options || {};
         Ext.Ajax.request({
-            url: this.urlprefix + this.getType() + "/get-data-by-id",
+            url: Routing.getBaseUrl() + "/admin/" + this.getType() + "/get-data-by-id",
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Resources/public/js/pimcore/document/document.js:20]: cause that is a base "controller", how do I get the route for page, document, snippet, etc?

File: bundles/AdminBundle/Controller/Admin/RecyclebinController.php [hunk 9]
```diff
@@ -27,7 +27,7 @@
 class RecyclebinController extends AdminController implements EventedControllerInterface
 {
     /**
-     * @Route("/recyclebin/list", methods={"POST"})
+     * @Route("/recyclebin/list", name="pimcore_admin_recyclebin_list", methods={"POST"})
```
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/RecyclebinController.php:30]: same as with others

**Comment by dpfaffenbauer** [comment]
@dvesh3 done. I don't think it is a good idea to change route names YET, also casue it is a BC break. But, you want it, so I did it....

**Comment by dvesh3** [comment]
Hey @dpfaffenbauer just confirmed with the team that this PR will most probably go into 6.7.0 release so you are right, it should not have BC breaks. Can you please revert route name changes? really sorry, i should have considered it before.

File: bundles/AdminBundle/Controller/Admin/ElementController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (dvesh3) [reply bundles/AdminBundle/Controller/Admin/ElementController.php:37]: It's just a nice-to-have but yes let's keep only required changes in this PR.

File: bundles/AdminBundle/Resources/public/js/pimcore/object/tags/hotspotimage.js [hunk 10]
```diff
@@ -62,17 +62,22 @@ pimcore.object.tags.hotspotimage = Class.create(pimcore.object.tags.image, {
...
-                            width: 88,
-                            height: 88,
-                            frame: true
-                        };
+                        params['width'] = 88;
+                        params['height'] = 88;
+                        params['frame'] = true;
 
-                        var url = Ext.String.urlAppend(baseUrl, Ext.Object.toQueryString(params));
+                        var url = Routing.generate('pimcore_admin_asset_getimagethumbnail', params);
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/tags/hotspotimage.js:80]: hotspotimage broken on grid view. "" opening tag missing, please see https://github.com/pimcore/pimcore/blob/f412ea8594c0d95ff38007af03b9f469a5b5a577/bundles/AdminBundle/Resources/public/js/pimcore/object/tags/hotspotimage.js#L87

File: bundles/AdminBundle/Resources/public/js/pimcore/workflow/transitions.js [hunk 11]
```diff
@@ -17,7 +17,7 @@ pimcore.workflow.transitions.perform = function (ctype, cid, elementEditor, work
 
 
     Ext.Ajax.request({
-        url :  transition.isGlobalAction ? '/admin/workflow/submit-global-action' : '/admin/workflow/submit-workflow-transition',
+        url : this.transitionConfig.isGlobalAction ? Routing.generate('pimcore_admin_workflow_submitglobal') : Routing.generate('pimcore_admin_workflow_submitworkflowtransition'),
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/public/js/pimcore/workflow/transitions.js:20]: `this.transitionConfig` not defined

File: bundles/AdminBundle/Resources/views/Admin/Asset/imageEditor.html.php [hunk 12]
```diff
@@ -72,7 +75,10 @@ class="color_area"
     $imageFileExtension = \Pimcore\File::getFileExtension($this->asset->getFilename());
     $imageUrl = $this->asset->getFullPath();
     if(!in_array($imageFileExtension, ['png', 'jpg', 'jpeg'])) {
-        $imageUrl = '/admin/asset/get-image-thumbnail?id=' . $this->asset->getId() . '&format=png';
+        $imageUrl = $view->router()->path('pimcore_admin_asset_getimagethumbnail', [
+            'id' => $this->asset->get(),
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/views/Admin/Asset/imageEditor.html.php:79]:
```suggestion
            'id' => $this->asset->getId(),
```

File: bundles/AdminBundle/Resources/views/SearchAdmin/Search/Quicksearch/asset.html.php [hunk 13]
```diff
@@ -7,13 +7,28 @@
...
     $previewImage = null;
     try {
-        $suffix = '&hdpi=true';
         if ($element instanceof \Pimcore\Model\Asset\Image) {
-            $previewImage = '/admin/asset/get-image-thumbnail?id=' . $element->getId() . '&treepreview=true' . $suffix;
-        } elseif ($element instanceof \Pimcore\Model\Asset\Video && \Pimcore\Video::isAvailable()) {
-            $previewImage = '/admin/asset/get-video-thumbnail?id=' . $element->getId() . '&treepreview=true'. $suffix;
-        } elseif ($element instanceof \Pimcore\Model\Asset\Document && \Pimcore\Document::isAvailable()) {
-            $previewImage = '/admin/asset/get-document-thumbnail?id=' . $element->getId() . '&treepreview=true' . $suffix;
+             $previewImage =$this->generateUrl('pimcore_admin_asset_getimagethumbnail', [
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/views/SearchAdmin/Search/Quicksearch/asset.html.php:11]: please simplify repeated url generation code as only route name changes based on conditions.

**Comment by dpfaffenbauer** [comment]
all done

File: bundles/AdminBundle/Resources/views/SearchAdmin/Search/Quicksearch/asset.html.php [hunk 14]
```diff
@@ -6,14 +6,23 @@
...
+
     try {
-        $suffix = '&hdpi=true';
         if ($element instanceof \Pimcore\Model\Asset\Image) {
-            $previewImage = '/admin/asset/get-image-thumbnail?id=' . $element->getId() . '&treepreview=true' . $suffix;
-        } elseif ($element instanceof \Pimcore\Model\Asset\Video && \Pimcore\Video::isAvailable()) {
-            $previewImage = '/admin/asset/get-video-thumbnail?id=' . $element->getId() . '&treepreview=true'. $suffix;
-        } elseif ($element instanceof \Pimcore\Model\Asset\Document && \Pimcore\Document::isAvailable()) {
-            $previewImage = '/admin/asset/get-document-thumbnail?id=' . $element->getId() . '&treepreview=true' . $suffix;
+             $previewImage =$this->generateUrl('pimcore_admin_asset_getimagethumbnail', $params);
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/views/SearchAdmin/Search/Quicksearch/asset.html.php:17]: Call to undefined method generateUrl

File: bundles/AdminBundle/Resources/public/js/pimcore/object/bulk-export.js [hunk 15]
```diff
@@ -22,7 +22,7 @@ pimcore.object.bulkexport = Class.create(pimcore.object.bulkbase, {
     export: function() {
 
         Ext.Ajax.request({
-            url: "/admin/class/bulk-export",
+            url: Routing.generate('pimcore_admin_dataobject_class_bulkexportprepare'),
```
[the remark this answers is not in the document]
Comment (dvesh3) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/bulk-export.js:25]: please fix route name to `pimcore_admin_dataobject_class_bulkexport`

File: bundles/AdminBundle/Resources/public/js/pimcore/object/classificationstore/collectionsPanel.js [hunk 16]
```diff
@@ -53,7 +53,7 @@ pimcore.object.classificationstore.collectionsPanel = Class.create({
             readerFields.push(columnConfig);
         }
 
-        var url = "/admin/classificationstore/collection-relations?";
+        var url = Routing.generate('pimcore_admin_dataobject_classificationstore_collectionrelations');
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/classificationstore/collectionsPanel.js:56]: No route found for "GET /admin/classificationstore/collection-relationsxaction=read"

Comment (dpfaffenbauer) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/classificationstore/collectionsPanel.js]: done

File: bundles/AdminBundle/Resources/views/SearchAdmin/Search/Quicksearch/asset.html.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Resources/views/SearchAdmin/Search/Quicksearch/asset.html.php]: done

**Comment by macghriogair** [comment]
I think it could be worth mentioning in the Upgrade Notes that this feature will break any existing overrides of core Pimcore routes via routing.yml unless the config is updated by setting `options.expose=true`. 

_Example_: 

in my project `routing.yml`

```yaml
pimcore_admin_dataobject_dataobject_treegetchildsbyid:
  path: /admin/object/tree-get-childs-by-id
  controller: AppBundle\Extensions\Controller\Admin\DataObjectController::myImprovedAction
  options:
    expose: true  # <- required for detection by FosJSRoutingBundle
```
(note that overriding the route via annotation won't work because there is no priority option yet).

Of course, one could argue that doing this is an edge case & not officially documented by Pimcore 😄

**Comment by dpfaffenbauer** [comment]
> Of course, one could argue that doing this is an edge case & not officially documented by Pimcore 😄

No, actually that is supported, and yes I agree. It should be documented.

**Comment by wujku** [comment]
Two routes does not exists after upgrade from 6.6.3 to 6.6.7:
- `pimcore_admin_dataobject_dataobject_treegetchildsbyid`
- `pimcore_admin_dataobject_dataobjecthelper_getbatchjobs`

**Comment by t-zilla** [comment]
> Two routes does not exists after upgrade from 6.6.3 to 6.6.7

I have the same issue in 6.6.8.
Screenshot_20200706_115412

**Comment by wujku** [comment]
@t-zilla are you overwriting / extending Pimcore controllers?

I had a problem with previously overwritten `getDefaultKeyBindingsAction` method from Pimcore `AdminController`.

```php
<?php

namespace AppBundle\Controller;

use AppBundle\Model\User;
use Pimcore\Bundle\AdminBundle\Controller\AdminController as BaseController;

class AdminController extends BaseController
{
    public function getDefaultKeyBindingsAction()
    {
        $data = User::getDefaultKeyBindings();

        return $this->adminJson(['success' => true, 'data' => $data]);
    }
}
```

Fixed by using `PimcoreModBundle` created by @dpfaffenbauer to replace core controller.
More info: https://talk.pimcore.org/t/how-to-override-admin-controller-or-action/1267/15

`app/config/config.yml`
```yml
pimcore_mod:
    admin_controllers:
        # /admin/user/get-default-key-bindings
        Pimcore\Bundle\AdminBundle\Controller\Admin\UserController: AppBundle\Controller\Admin\UserController
```

`src/AppBundle/Resources/config/pimcore/routing.yml`
```diff
- app:
-     resource: "@AppBundle/Controller/"
-     type:     annotation
```

`src/AppBundle/Controller/Admin/UserController.php`
```php
<?php

namespace AppBundle\Controller\Admin;

use AppBundle\Model\User;
use Pimcore\Bundle\AdminBundle\Controller\Admin\UserController as BaseController;
use Pimcore\Bundle\AdminBundle\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Annotation\Route;

class UserController extends BaseController
{
    /**
     * @Route("/user/get-default-key-bindings", name="pimcore_admin_user_getdefaultkeybindings", methods={"GET"})
     *
     * @param Request $request
     *
     * @return JsonResponse
     */
    public function getDefaultKeyBindingsAction(Request $request)
    {
        $data = User::getDefaultKeyBindings();

        return $this->adminJson(['success' => true, 'data' => $data]);
    }
}
```

**Comment by zeor59** [comment]
same problem after upgrade from 6.6.4 to 6.6.9

image

That's works when I delete my "DataObjectHelperController.php" which extends Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectHelperController

Can someone helps me please ?

my routing.yml file is: 

app:
    resource: "@AppBundle/Controller/"
    type:     annotation

Thanks in advance for your help

**Comment by fashxp** [comment]
please create a new issue, thx.

**Comment by wujku** [comment]
@zeor59 which methods are you trying to overwrite?

**Comment by zeor59** [comment]
> 
> 
> @zeor59 which methods are you trying to overwrite?

Sorry for my late answer.
My problem was fixed with adding

```
  options:
    expose: true

to my routing.yml file.


Can someone explains me why this option is now necessary ?
```

Thanks

**Comment by dpfaffenbauer** [comment]
@zeor59 cause routes in Js are now generated using JavaScript (FosRoutingBundle). If you overwrite routes, you have to expose them to FosRouting.

**Comment by zeor59** [comment]
> 
> 
> @zeor59 cause routes in Js are now generated using JavaScript (FosRoutingBundle). If you overwrite routes, you have to expose them to FosRouting.

Thanks you

**Comment by nicklassandell** [comment]
Hi y'all. I'm having some issues with editmode in all documents being broken after updating. After digging around a bit it looks like the problem occurs when assigning an asset version in the framework config. It leads to file_get_contents() failing at one point because a query param is added to a file path. I have the following defined in my config, and it simply adds ?xxx to any asset URL, and the ASSET_CACHE_VERSION is read from an .env file which is re-generated on deployment.

```
framework:
    assets:
        version: '%env(ASSET_CACHE_VERSION)%'
```

This line here appears to be the root cause. It works fine if you just remove the $this->package->getUrl() and add a slash in the beginning. @dpfaffenbauer is there a reason for the getUrl call? https://github.com/pimcore/pimcore/blob/master/bundles/CoreBundle/EventListener/Frontend/EditmodeListener.php#L321

The resulting file path is something like this, which file_get_contents can't handle.

`/var/www/html/web/bundles/fosjsrouting/js/router.js?23471` <-- see query param

Here's where the file_get_contents() happens:
https://github.com/pimcore/pimcore/blob/master/bundles/CoreBundle/EventListener/Frontend/EditmodeListener.php#L274

**Comment by dpfaffenbauer** [comment]
@nicklassandell I have no idea anymore why I used getUrl for this....

**Pull request merged**